### PR TITLE
Updated SetOperationCornerCases for Unions and cleaned up the code.

### PR DESCRIPTION
**Pull request by leerho**

### CODE REVIEW

File: src/main/java/org/apache/datasketches/tuple/Union.java [hunk 1]
```diff
@@ -134,64 +130,78 @@ public void union(final org.apache.datasketches.theta.Sketch thetaSketch, final
...
-    if (thetaLong_ >= qsk_.thetaLong_ && qsk_.getRetainedEntries() <= qsk_.getNominalEntries()) {
-      return qsk_.compact();
-    }
-    long theta = min(thetaLong_, qsk_.thetaLong_);
+      result = qsk_.compact();
+    } else if (thetaLong_ >= qsk_.thetaLong_ && qsk_.getRetainedEntries() <= qsk_.getNominalEntries()) {
+      result = qsk_.compact();
+    } else {
+      long theta = min(thetaLong_, qsk_.thetaLong_);
+      int numHashes = 0;
```
Comment (davecromberge) [inline src/main/java/org/apache/datasketches/tuple/Union.java:167]: Could you please explain the second "if" branch with regard to the state combination truth table?  From some of the conditional checks, it is clear that `qsk` is an exact mode sketch.  It is also clear from the first branch that the local state is not empty.  Where the difficulty arises is that the state for `qsk` and `thetaLong` is non-local to this function and requires more context.

**Review by davecromberge — APPROVED** [review]

File: src/main/java/org/apache/datasketches/tuple/Union.java (hunk 1 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/Union.java:167]: This is confusingly written, I admit. I am looking into how it can be made more direct.  While studying this, I may have found another problem, which I need to investigate.

Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/Union.java]: I have cleaned up this section and added comments so that I could figure out what exactly was going on :)

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java [hunk 2]
```diff
@@ -20,6 +20,8 @@
 package org.apache.datasketches.theta;
 
 import org.testng.annotations.Test;
+//import static org.apache.datasketches.Util.DEFAULT_UPDATE_SEED;
+//import static org.apache.datasketches.hash.MurmurHash3.hash;
```
Comment (AlexanderSaydakov) [inline src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:25]: do we need to keep these commented-out lines?

File: src/main/java/org/apache/datasketches/tuple/Union.java (hunk 1 above)
Comment (davecromberge) [reply src/main/java/org/apache/datasketches/tuple/Union.java]: The additional comments were helpful and it makes more sense now - thanks!

**Comment by leerho** [comment]
I'm going to add to this the corner-case tests for the ArrayOfDoubles Tuple Sketch.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java (hunk 2 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:25]: These two imports are associated with some helper methods at the bottom of the file, which are also commented out.  These helper methods would aid a reader in understanding how the various test hash values used in all the tests are actually generated.

**Comment by leerho** [comment]
Lots of changes in this PR.  A lot of changes in var names to make the code more understandable.  Also lots of changes in javadocs and code comments.  But the code runs all the regular tests plus all of the corner case tests for theta, tuple (all the generics) and the ArrayOfDoubles Tuple sketches.  Code coverage is still above 96%. So unless something comes up in the review, it should be ready for a release.  

I did discover a problem with Backward compatibility with SerVer 1 theta sketches that works (in a strange way) internally in Yahoo, but was not done correctly.  Once Will has moved his code to V3, we can deprecate all the compatibility with V1 and V2.  These old Serialization Versions have never appeared in Open Source code.  We should be able to make this change sometime in Q1/2022. 

Please Review, 

Thanks

**Review by AlexanderSaydakov — APPROVED** [review]
